### PR TITLE
feat: add moving tab to other position

### PR DESCRIPTION
**Pull request by devzbysiu**
### Details:
- when we want to move the first tab to the left or the last tab to the right, the **implementation wraps the tabs around** (i.e.: moving first tab left will make this tab the last one, moving last tab right will make it the first one), so it switches the tab with the tab on the left/right as if the tabs are in circular buffer. This implementation allows changing tabs without the need to move all of the other tabs to different position. See the demo below.
- at this moment, the key bindings are set (in normal mode) to `Alt + i` for moving tab left and `Alt + o` for moving tab right (no particular reason for those bindings, I just picked something available and close to "`hjkl` row")
- major changes are:
  - ~~changelog~~ - changed separately after merge
  - added tip for this feature, here I also did small refactor of the tip files (moved the repeated `strings!` macro to the main module file and reused it in all tip files)
  - e2e tests, here I  moved the steps definition to separate file so as not to clutter the cases file + snapshots
  - the feature itself (protobuf changes for new action and direction of move, handling the action) + unit tests

Demo:
- first I'm changing the tab name
- then showing the contents of all tabs
- then moving the tab around
- again showing contents of all tabs
- again some moving around

[zellij-tabs-showcase.webm](https://github.com/zellij-org/zellij/assets/50180524/9fcf17fb-9b51-4f8d-b4b7-f6349169621e)

_Edit:_
Demo of more complex scenarios:

[zellij-tabs-complex-showcase.webm](https://github.com/zellij-org/zellij/assets/50180524/285cfe60-3f25-4bab-99f3-6521bfdaa8da)

### CODE REVIEW

**Comment by Andrew15-5** [comment]
I think a much visually impressive demo would be if each tab had multiple panes, this also would serve as a test/reassurance that this feature does work with multi-pane tabs. Or maybe some tabs have 1 pane, and some tabs have multiple panes.

**Comment by devzbysiu** [comment]
> I think a much visually impressive demo would be if each tab had multiple panes, this also would serve as a test/reassurance that this feature does work with multi-pane tabs. Or maybe some tabs have 1 pane, and some tabs have multiple panes.

That's a good point. Added demo with more complex scenarios (multi-pane, single-pane, sync tabs).

**Comment by jaeheonji** [comment]
It really looks cool! I'll review it soon :)

Before that, I have one request. Please remove the `CHANGELOG.md` in changes. I will apply this after merging the PR seperatly.

**Comment by devzbysiu** [comment]
Hi @jaeheonji, thank you :)

I'll revert `CHANGELOG.md` asap. I'll also fix the tests (my bad, I focused on e2e and missed `cargo xtask test` :man_facepalming: )

**Comment by mmirus** [comment]
Thanks for working on this feature! I'm a new user and was missing it.

File: zellij-utils/src/data.rs [hunk 1]
```diff
@@ -251,6 +251,35 @@ impl FromStr for Direction {
     }
 }
 
+#[derive(Eq, Clone, Copy, Debug, PartialEq, Hash, Deserialize, Serialize, PartialOrd, Ord)]
+pub enum HorizontalDirection {
```
Comment (jaeheonji) [inline zellij-utils/src/data.rs:255]: Can't we re-use an already existing `Direction` enum?

When the actual API is called, the `ProtobufMoveTabDirection` data appears to be mapped to a `HorizontalDirection`, so it looks like the existing `Direction` can be used. (If you handle an exception in the vertical case)

What do you think?

File: zellij-utils/assets/prost/api.action.rs [hunk 2]
```diff
@@ -587,6 +591,33 @@ impl ActionName {
...
             "RenameSession" => Some(Self::RenameSession),
             "LaunchPlugin" => Some(Self::LaunchPlugin),
+            "MoveTab" => Some(Self::MoveTab),
+            _ => None,
+        }
+    }
+}
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, ::prost::Enumeration)]
+#[repr(i32)]
+pub enum MoveTabDirection {
```
Comment (jaeheonji) [inline zellij-utils/assets/prost/api.action.rs:601]: Can you move this to line 329? I think it would be better for visibility if Enum are placed close to each other.

File: zellij-server/src/screen.rs [hunk 3]
```diff
@@ -1530,6 +1535,77 @@ impl Screen {
         }
     }
 
+    pub fn move_active_tab_to_left(&mut self, client_id: ClientId) -> Result<()> {
+        if self.tabs.len() < 2 {
+            info!("cannot move tab to left: only one tab exists");
```
Comment (jaeheonji) [inline zellij-server/src/screen.rs:1540]: I think it would be better to do it with `debug`.

File: zellij-server/src/screen.rs [hunk 4]
```diff
@@ -1530,6 +1535,77 @@ impl Screen {
...
+        //            current_index               index_on_left_of(current_index)
+        let tabs_count = self.tabs.len();
+        (current_index + tabs_count - 1) % tabs_count
+    }
+
+    fn switch_tabs(&mut self, active_tab_idx: usize, other_tab_idx: usize, client_id: u16) {
+        let (Some(mut active_tab), Some(mut other_tab)) = (
+            self.tabs.remove(&active_tab_idx),
+            self.tabs.remove(&other_tab_idx),
+        ) else {
```
Comment (jaeheonji) [inline zellij-server/src/screen.rs:1572]: I have a question in here.

For unknown reasons, I think if `active_tab` is `Some` and `other_tab` is `None`, this condition is not satisfied and the function just returns.

However, since `remove` has already been called, `active_tab` will disappear from tabs.

For safer code, it seems better to check whether the key exists first. What do you think?

File: default-plugins/status-bar/src/tip/data/mod.rs [hunk 5]
```diff
@@ -88,5 +89,25 @@ lazy_static! {
...
+macro_rules! ansi_strings {
+    ($ANSIStrings:expr) => {{
+        let strings: &[::ansi_term::ANSIString] = $ANSIStrings;
+        let ansi_strings = ::ansi_term::ANSIStrings(strings);
+        LinePart {
+            part: format!("{}", ansi_strings),
+            len: ::ansi_term::unstyled_len(&ansi_strings),
+        }
+    }};
+}
```
Comment (jaeheonji) [inline default-plugins/status-bar/src/tip/data/mod.rs:113]: It's a good refactoring, but could you leave this out of this PR and make it a separate PR?

(In my opinion, separating PR according to purpose makes management more convenient.)

**Review by jaeheonji — CHANGES_REQUESTED** [review]
I'm sorry. The review is late due to a personal issue. I haven't run the test yet, but I left a few comments. 😃

File: zellij-utils/src/data.rs (hunk 1 above)
Comment (devzbysiu) [reply zellij-utils/src/data.rs]: Yeah, I was thinking about that. I figured that there already is `SearchDirection`, `ResizeDirection`, `SpliDirection` etc., so I introduced `HorizontalDirection`, but I don't mind reusing `Direction`, it makes sense.

Please just make sure I handled the [vertical case exception](https://github.com/zellij-org/zellij/pull/3047/files#diff-6f77a56f2f7ebfaa89005876b659af0aa348b809cf0eebda1814e1c7454a637cR541) in a proper way, because I'm not entirely sure :thinking:

File: zellij-utils/assets/prost/api.action.rs (hunk 2 above)
Comment (devzbysiu) [reply zellij-utils/assets/prost/api.action.rs]: Yep, done :+1:

File: zellij-server/src/screen.rs (hunk 3 above)
Comment (devzbysiu) [reply zellij-server/src/screen.rs]: Done :+1:

File: zellij-server/src/screen.rs (hunk 4 above)
Comment (devzbysiu) [reply zellij-server/src/screen.rs]: That's a good catch. In this case it should be guaranteed that both `active_tab` and `other_tab` is `Some`, because the method `switch_tabs` is called only when:

1. There are at least two tabs.
2. `active_tab` is `Some`.

So only the `other_tab` thing left and it's guaranteed to be `Some` as well, because that's how `index_on_left_of` and `index_on_right_of` are implemented.

But anyway - who knows how the code will evolve, so I agree that more defensive approach is better. PR updated :+1:

File: default-plugins/status-bar/src/tip/data/mod.rs (hunk 5 above)
Comment (devzbysiu) [reply default-plugins/status-bar/src/tip/data/mod.rs]: Yeah, I was thinking about that as well. I'll move the changes to separate PR :+1:

Comment (devzbysiu) [reply default-plugins/status-bar/src/tip/data/mod.rs]: Moved [here](https://github.com/zellij-org/zellij/pull/3086)

**Review by jaeheonji — CHANGES_REQUESTED** [review]
Great! I think we are almost done.

Functionally I think I'm ready to merge now. But before that, I suggest some minor refactoring. Please take a look.

**Comment by devzbysiu** [comment]
@jaeheonji perfect. Thank you for the review :+1: . I'll take a look asap. Probably this weekend

**Comment by jaeheonji** [comment]
@devzbysiu Great :)

Actually, We plan to make a new release this weekend or next week. And we're hoping to include this feature in the next version. So, if you need any help, please let me know! then, I will update the refactoring part.

File: zellij-utils/src/kdl/mod.rs [hunk 6]
```diff
@@ -453,6 +453,16 @@ impl Action {
...
                 Ok(Action::MoveFocusOrTab(direction))
             },
+            "MoveTab" => {
+                let direction = Direction::from_str(string.as_str()).map_err(|_| {
+                    ConfigError::new_kdl_error(
+                        format!("Invalid horizontal direction: '{}'", string),
+                        action_node.span().offset(),
+                        action_node.span().len(),
+                    )
+                })?;
```
[the remark this answers is not in the document]
Comment (devzbysiu) [reply zellij-utils/src/kdl/mod.rs]: Oh, yeah I missed that. Done

File: zellij-server/src/screen.rs [hunk 7]
```diff
@@ -1537,6 +1542,97 @@ impl Screen {
...
+            return Ok(());
+        };
+        let Some(&active_tab_idx) = self.active_tab_indices.get(&client_id) else {
+            return Ok(());
+        };
+        let left_tab_idx = self.index_on_left_of(active_tab_idx);
+        self.switch_tabs(active_tab_idx, left_tab_idx, client_id);
+        self.log_and_report_session_state()
+            .context("failed to move tab to left")?;
+        Ok(())
```
[the remark this answers is not in the document]
Comment (devzbysiu) [reply zellij-server/src/screen.rs]: Not sure if I get what you mean. But if I got it right, then I can't do that this way because:

- to calculate target index in step 1. I need `active_tab_idx`, so I'm checking it before
- to get `active_tab_idx` in the first place, via `self_active_tab_indices.get(&client_id)`, I need `client_id`, so I'm checking it before

To make the checks of `client_id` and `active_tab_index` DRY I can make `move_active_tab(client_id: ClientID, direction: Direction)` , get rid of `move_active_tab_right` and `move_active_tab_left` and then based on the `direction` passed in, calculate the target tab index. Something like this:

```
pub fn move_active_tab(&mut self, client_id: ClientId, dir: Direction) -> Result<()> {
        if dir.is_vertical() {
            panic!("cannot move tab vertically");
        }
        if self.tabs.len() < 2 {
            debug!("cannot move tab: only one tab exists");
            return Ok(());
        }
        let Some(client_id) = self.client_id(client_id) else {
            return Ok(());
        };
        let Some(&active_tab_idx) = self.active_tab_indices.get(&client_id) else {
            return Ok(());
        };

        let target_idx = if dir == Direction::Left {
            // wraps around: [tab1, tab2, tab3] => [tab1, tab2, tab3]
            //                 ^                                 ^
            //          active_tab_idx                     target_idx
            (active_tab_idx + self.tabs.len() - 1) % self.tabs.len()
        } else {
            // wraps around: [tab1, tab2, tab3] => [tab1, tab2, tab3]
            //                             ^          ^
            //                     active_tab_idx   target_idx
            (active_tab_idx + 1) % self.tabs.len()
        };

        self.switch_tabs(active_tab_idx, target_idx, client_id);
        self.log_and_report_session_state()
            .context("failed to move active tab")?;
        Ok(())
    }
```

[the remark this answers is not in the document]
Comment (jaeheonji) [reply zellij-server/src/screen.rs]: Actually, Yes I was thinking exactly this one. However, I think it would be better to postpone this refactoring until later.
Btw, Thank you for sharing your idea!

**Comment by jaeheonji** [comment]
LGTM. Really Good!

After CI is passed, I will merge as soon as possible.

**Review by jaeheonji — APPROVED** [review]

**Comment by janos-r** [comment]
Omg after years finally!
Now I assume 
https://github.com/zellij-org/zellij/issues/3080
https://github.com/zellij-org/zellij/issues/1656 and
https://github.com/zellij-org/zellij/issues/1654
can all be closed if the feature gets released and works

**Comment by DanielAndreasen** [comment]
Speaking of, when can we expect this feature in a new release?

**Comment by giyany** [comment]
It's highly likely this release happens in April, but please don't come after us if it takes longer.

**Comment by youcefs21** [comment]
Hi, I know you said to not come after you if it takes longer, but it is almost end of may now, and I'd love an update on the progress

**Comment by DanielAndreasen** [comment]
It's in the latest release here: https://github.com/zellij-org/zellij/releases/tag/v0.40.1
And it works wonderfully! :tada:

**Comment by boblehest** [comment]
> It's in the latest release here: https://github.com/zellij-org/zellij/releases/tag/v0.40.1 And it works wonderfully! 🎉

Is it described in the release notes? I can't see any mention of it.

**Comment by matteomanuelli** [comment]
> It's in the latest release here: https://github.com/zellij-org/zellij/releases/tag/v0.40.1 And it works wonderfully! 🎉

~I can't find~ we can't find the keybinding to move tab to another position

**Comment by Andrew15-5** [comment]
Same bindings as in the OP. You probably can get a new version of the config by renaming the existing config file. I think they should appear there.

**Comment by pdeva** [comment]
doing `option-o`/`option-i` doesnt work in macos...

**Comment by danieldk** [comment]
> doing `option-o`/`option-i` doesnt work in macos...

Yeah, these key combinations are used by default on macOS for the slanted o/caret. I added these key bindings, so that I can do Shift left/right in tab mode to move around tabs:

```
keybinds {
        tab {
                bind "Shift Left" {
                        MoveTab "Left"
                }
                bind "Shift Right" {
                        MoveTab "Right"
                }
        }
}
```

**Comment by ailequal** [comment]
> > doing `option-o`/`option-i` doesnt work in macos...
> 
> Yeah, these key combinations are used by default on macOS for the slanted o/caret. I added these key bindings, so that I can do Shift left/right in tab mode to move around tabs:
> 
> ```
> keybinds {
>         tab {
>                 bind "Shift Left" {
>                         MoveTab "Left"
>                 }
>                 bind "Shift Right" {
>                         MoveTab "Right"
>                 }
>         }
> }
> ```

Nice idea, love it! I improve it a little bit by also adding the "h" and "l" for moving with shift.

```
        bind "Shift Left" { MoveTab "Left"; }
        bind "Shift h" { MoveTab "Left"; }
        bind "Shift Right" { MoveTab "Right"; }
        bind "Shift l" { MoveTab "Right"; }
```

**Comment by Johnsoct** [comment]
Love the feature, thank you for adding it, but the "alt/option + [io]" keybind is missing from the default keybind documentation: https://github.com/zellij-org/zellij/blob/main/zellij-utils/assets/config/default.kdl

**Comment by slevi123** [comment]
Hey everyone, I would appreciate if the default keybindings help panel would contain the help message for this tab moving feature. I mean in the compact mode floating panel. 
(I am new to Zellij, I do not really know the bits of it)

Can you point me towards a cleaner definition of what I want?

**Comment by Andrew15-5** [comment]
Hi. Please open a new issue about this instead of asking here. Or open a question in discussions tab.
